Move out `tendermint::config` to `tendermint-config` crate

soareschen opened the pull request:
Fixes #983.

* [ ] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Added entry in `.changelog/`

xla reviewed (commented):
The metadata in the new cargo toml should be catered to the actual crate instead of a verbatim copy of the tendermint cargo manifest.

soareschen commented:
> The metadata in the new cargo toml should be catered to the actual crate instead of a verbatim copy of the tendermint cargo manifest.

Yeah I missed updating that after copying the boilerplate. Fixed it now.

soareschen commented:
Rehashing my note in #983:

I have taken a look at the use of `net::Address`, and it does not seem like the actual details are used anywhere. In #986 I moved out `tendermint::net` to `tendermint_config::net`, and changed the field in `tendermint::node::OtherInfo::rpc_address` to just `String`. If anyone needs to do the actual URL parsing, they can do it at the call site instead.

The main difference I see of using `String` instead of `Address` for `rpc_address` is the way it is deserialized. With the `Deserialize` implementation provided by `Address`, an invalid URL would result in an error. However I also do not see how this can be solved using feature flag. In the absense of the `url` crate, should the deserialization just succeed? That wouldn't be good since the crate would effectively be behaving differently with no invariant in place. Otherwise if it is possible to implement the parsing without using `url`, then it is better off to just not `url` in both cases in the first place.

Note that the other alternative is to just wait for servo/rust-url#717 to get merged and released.

soareschen commented on light-client/src/components/verifier.rs:
```diff
@@ -128,14 +128,12 @@ where
-    /// - Otherwise, ensure that the untrusted block has a greater height than
-    ///   the trusted block.
+    /// - If the untrusted block is the very next block after the trusted block, check that their
+    ///   (next) validator sets hashes match.
```
Not sure about that. This is somehow modified by `cargo fmt` even though no code changes are made.

soareschen commented on light-client/src/components/verifier.rs:
```diff
@@ -128,14 +128,12 @@ where
     /// - Ensure the header validator hashes match the given validators
-    /// - Ensure the header next validator hashes match the given next
-    ///   validators
+    /// - Ensure the header next validator hashes match the given next validators
```
Not sure about that. This is somehow modified by `cargo fmt` even though no code changes are made.

seanchen1991 replied:
Oh, I meant the content of the actual comment, not the formatting that got changed (though to be fair, I probably wouldn't have noticed this comment if it hadn't gotten re-formatted) 🙂

soareschen replied:
Yeah, I think you can ask in #tendermint on Slack to clarify.

soareschen commented on config/src/net.rs:
```diff
@@ -83,7 +96,8 @@ impl FromStr for Address {
             "tcp" => Ok(Self::Tcp {
                 peer_id: if !url.username().is_empty() {
-                    Some(url.username().parse()?)
+                    let username = url.username().parse().map_err(Error::tendermint)?;
```
A new `Error` type is defined for `tendermint-config`. The error constructor `Error::tendermint` wraps a `tendermint::Error` into a `tendermint_config::Error`.

soareschen commented on tendermint/src/lib.rs:
```diff
@@ -5,6 +5,7 @@
 //! and remote procedure calls (JSON-RPC).
 
 #![cfg_attr(docsrs, feature(doc_cfg))]
+#![allow(dead_code)]
```
Stale changes. Removed.

thanethomson commented on config/Cargo.toml:
```diff
@@ -0,0 +1,49 @@
+[package]
+name       = "tendermint-config"
+version    = "0.21.0" # Also update `html_root_url` in lib.rs and
```
```suggestion
version    = "0.22.0" # Also update `html_root_url` in lib.rs and
```

thanethomson commented on config/Cargo.toml:
```diff
@@ -0,0 +1,49 @@
+
+authors = [
+    "Informal Systems <hello@informal.systems>",
+    "Ismail Khoffi <Ismail.Khoffi@gmail.com>",
+    "ValarDragon <dojha12@gmail.com>",
+    "Tony Arcieri <tony@iqlusion.io>",
```
```suggestion
```

thanethomson commented on rpc/Cargo.toml:
```diff
@@ -74,6 +74,7 @@ serde = { version = "1", features = [ "derive" ] }
 serde_bytes = "0.11"
 serde_json = "1"
 tendermint = { version = "0.21.0", path = "../tendermint" }
+tendermint-config = { version = "0.21.0", path = "../config" }
```
```suggestion
tendermint-config = { version = "0.22.0", path = "../config" }
```

thanethomson commented on config/tests/mod.rs:
```diff
@@ -0,0 +1,239 @@
+//!
+//! Test config files are located in the `tests/support/config` subdirectory.
+
+mod files {
```
Is this file not the same as `config/tests/config.rs`?

thanethomson replied:
Also, what's the benefit of wrapping these tests in the `files` mod? Could we not just do away with the outer `mod files` wrapper?

soareschen replied:
> Is this file not the same as `config/tests/config.rs`?

Oh yeah, that was probably a mistake during the file moving. I have removed the duplicate.


> Also, what's the benefit of wrapping these tests in the `files` mod? Could we not just do away with the outer `mod files` wrapper?

Yes we can. The `mod files` was there in the original code. I have moved them out now. One disadvantage is that git now recognize the changes as a new file rather than a moved file.

thanethomson replied:
> Yes we can. The mod files was there in the original code. I have moved them out now. One disadvantage is that git now recognize the changes as a new file rather than a moved file.

That's fine, thanks!

thanethomson commented on abci/src/lib.rs:
````diff
@@ -9,7 +9,7 @@
 //! by a full Tendermint node.
 //!
-//! ```rust
+//! ```ignore
````
Was this doctest failing?

soareschen replied:
Yup. `cargo test` on the project root fails because of this and a few other doctests.

thanethomson replied:
I've submitted #990 to fix these problems.

xla replied:
It would be preferrable to not just ignore parts that break. The least would have been a tracking issue, a fix like @thanethomson provided is obviously preferrable.

soareschen replied:
Nice! I will revert this change then.

soareschen commented on tendermint/src/node/info.rs:
```diff
@@ -82,7 +76,7 @@ pub struct OtherInfo {
 
     /// RPC address
-    pub rpc_address: net::Address,
+    pub rpc_address: String,
```
One unresolved issue is whether using `String` instead of `net::Address` for `rpc_address` will break any functionality. The main difference is in the implementations of `Deserialize` and `Display`. Any idea of how this field is used in practice?

xla replied:
@thanethomson Might have some input here.

thanethomson replied:
No idea here actually.
